File: ruby_practice_problems/02_netflix/fetch_top_movies.rb
require './LinkedList.rb'

# O(n * k^2) - O(nk) - max list length * k number of lists
# Space O(1)

def merge2_country(l1, l2)
    dummy = Node.new(-1)
    curr = dummy

    while l1 and l2
        if l1.value <= l2.value
            curr.next = l1
            l1 = l1.next
        else
            curr.next = l2
            l2 = l2.next
        end
        curr = curr.next
    end

    if l1 != nil
        curr.next = l1
    else
        curr.next = l2
    end
    return dummy.next
end

def mergeK_country(lists)
    if lists.length() > 0
        res = lists[0]
        (1..lists.length() - 1).each do |i|
            res = merge2_country(res, lists[i])
        end
        return res
    end
    return nil
end

a = create_linked_list([11,41,51])
b = create_linked_list([21,23,42])
c = create_linked_list([25,56,66,72])

puts "Result:"
display(mergeK_country([a, b, c]))

File: javascript_practice_problems/01_netflix/fetch_top_movies.js
import {LinkedListNode, createLinkedList, toList} from './LinkedList.js'


let merge2Country = function(head1, head2) {
    // if both lists are empty then merged list is also empty
    // if one of the lists is empty then other is the merged list
    if (!head1) {
      return head2;
    } else if (!head2) {
      return head1;
    }

    let mergedHead = null;
    if (head1.data <= head2.data) {
      mergedHead = head1;
      head1 = head1.next;
    } else {
      mergedHead = head2;
      head2 = head2.next;
    }

    let curr = mergedHead;

    while (head1 && head2) {
      let temp = null;
      if (head1.data <= head2.data) {
        temp = head1;
        head1 = head1.next;
      } else {
        temp = head2;
        head2 = head2.next;
      }
      curr.next = temp;
      curr = temp;
    }
    if (head1) {
      curr.next = head1;
    } else if (head2) {
      curr.next = head2;
    }

    return mergedHead;
};

function mergeKCounty(lists){ // Main function

    if (lists.length > 0){
        res = lists[0];
        for (var i = 1; i < lists.length; i++) {
            res = merge2Country(res, lists[i]);
        }
        return res;
    }
}